### ISSUE TITLE
Java 11 NoSuchAlgorithmException SHA1PRNG when FIPS enabled TS012071744

### ISSUE THREAD

**Issue by barbj** [OpenLiberty/open-liberty#24469]
### Describe the bug

When enabling FIPS feature with the IBM Semeru JDK and OpenID connect client enabled this  ffdc exception is emitted: 

```
------Start of DE processing------ = [2/3/23, 18:08:06:748 CET]
Exception = java.security.NoSuchAlgorithmException
Source = com.ibm.ws.security.openidconnect.clients.common.OidcUtil
probeid = 104
Stack Dump = java.security.NoSuchAlgorithmException: SHA1PRNG SecureRandom not available
        at java.base/sun.security.jca.GetInstance.getInstance(GetInstance.java:159)
        at java.base/java.security.SecureRandom.getInstance(SecureRandom.java:398)
        at com.ibm.ws.security.openidconnect.clients.common.OidcUtil.getRandom(OidcUtil.java:102)
        at com.ibm.ws.security.openidconnect.clients.common.OidcUtil.generateRandom(OidcUtil.java:82)
        at com.ibm.ws.security.openidconnect.client.internal.OidcClientConfigImpl.<clinit>(OidcClientConfigImpl.java:307)
        at com.ibm.ws.security.openidconnect.client.internal.OidcClientWebappConfigImpl.activate(OidcClientWebappConfigImpl.java:47)
        at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
```
### **Input from JDK support**

> When FIPS is enabled it certainly does change the algorithms that are available. The algorithms that are enabled are enabled for a number of reasons:
>
> 1. The libraries that are able to handle the algorithm are indeed FIPS certified. At this time this includes the NSS library on red hat.
> 2. The provider / algorithm does not provide cryptography that needs to be certified. This includes the SunJSSE provider.
>
>At this time we do not have a random number generator available that is FIPS certified so the algorithms you outline SHA1PRNG, NativePRNG & DRBG are unavailable.
>
>The algorithms are based upon the NSS library from red hat. When they certify the software only certain algorithms are allowed. We then go through and enable these algorithms in the Java layer.
>
>For more details please reference ,  FIPS certified cryptography in IBM Semeru JDK: https://www.ibm.com/support/pages/node/6612693 it talks generally about the NSS FIPS-2 certificate and has a link to the red hat certification registered at NIST for the NSS library.

### **Diagnostic information:**  
 - Liberty 22.0.0.13 (Reproduce also with 22.0.0.12)
 - Affected feature(s) 
   - openidConnectClient-1.0, transportSecurity-1.0
 - Java Version: 
   - java.version = 11.0.17
   - java.runtime = IBM Semeru Runtime Open Edition (11.0.17+8)
   - os = Linux (4.18.0-372.40.1.el8_6.x86_64; amd64) (en_US)
 - server.xml configuration 

```
<server>
<featureManager>
<feature>servlet-3.1</feature>
<feature>appSecurity-2.0</feature>
<feature>jsp-2.3</feature>
<feature>jdbc-4.1</feature>
<feature>ldapRegistry-3.0</feature>
<feature>openidConnectClient-1.0</feature>
<feature>transportSecurity-1.0</feature>
<feature>monitor-1.0</feature>
</featureManager>
</server>
<server>
<sslDefault sslRef="odmDefaultSSLConfig"/>
<ssl id="odmDefaultSSLConfig" keyStoreRef="odmDefaultKeyStore" sslProtocol="TLSv1.2" enabledCiphers="AES-256-GCM AES-256-CCM CHACHA20-POLY1305 AES-256-CBC AES-128-GCM AES-128-CCM AES-128-CBC"/>
<keyStore id="odmDefaultKeyStore" location="/config/security/volume/pkcs11cfg.cfg" type="PKCS11-NSS-FIPS" fileBased="false" password="changeit" provider="SunPKCS11-NSS-FIPS"/>
</server>
```
### **Full ffdc**
```
------Start of DE processing------ = [2/8/23, 13:55:53:391 CET]
Exception = java.security.NoSuchAlgorithmException
Source = com.ibm.ws.security.openidconnect.clients.common.OidcUtil
probeid = 104
Stack Dump = java.security.NoSuchAlgorithmException: SHA1PRNG SecureRandom not available
	at java.base/sun.security.jca.GetInstance.getInstance(GetInstance.java:159)
	at java.base/java.security.SecureRandom.getInstance(SecureRandom.java:398)
	at com.ibm.ws.security.openidconnect.clients.common.OidcUtil.getRandom(OidcUtil.java:102)
	at com.ibm.ws.security.openidconnect.clients.common.OidcUtil.generateRandom(OidcUtil.java:82)
	at com.ibm.ws.security.openidconnect.client.internal.OidcClientConfigImpl.<clinit>(OidcClientConfigImpl.java:307)
	at com.ibm.ws.security.openidconnect.client.internal.OidcClientWebappConfigImpl.activate(OidcClientWebappConfigImpl.java:47)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.base/java.lang.reflect.Method.invoke(Method.java:566)
	at org.apache.felix.scr.impl.inject.methods.BaseMethod.invokeMethod(BaseMethod.java:244)
	at org.apache.felix.scr.impl.inject.methods.BaseMethod.access$500(BaseMethod.java:41)
	at org.apache.felix.scr.impl.inject.methods.BaseMethod$Resolved.invoke(BaseMethod.java:685)
	at org.apache.felix.scr.impl.inject.methods.BaseMethod.invoke(BaseMethod.java:529)
	at org.apache.felix.scr.impl.inject.methods.ActivateMethod.invoke(ActivateMethod.java:318)
	at org.apache.felix.scr.impl.inject.methods.ActivateMethod.invoke(ActivateMethod.java:308)
	at org.apache.felix.scr.impl.manager.SingleComponentManager.createImplementationObject(SingleComponentManager.java:354)
	at org.apache.felix.scr.impl.manager.SingleComponentManager.createComponent(SingleComponentManager.java:115)
	at org.apache.felix.scr.impl.manager.SingleComponentManager.getService(SingleComponentManager.java:1000)
	at org.apache.felix.scr.impl.manager.SingleComponentManager.getServiceInternal(SingleComponentManager.java:973)
	at org.apache.felix.scr.impl.manager.AbstractComponentManager.activateInternal(AbstractComponentManager.java:776)
	at org.apache.felix.scr.impl.manager.AbstractComponentManager.enableInternal(AbstractComponentManager.java:674)
	at org.apache.felix.scr.impl.manager.AbstractComponentManager.enable(AbstractComponentManager.java:437)
	at org.apache.felix.scr.impl.manager.ConfigurableComponentHolder.enableComponents(ConfigurableComponentHolder.java:667)
	at org.apache.felix.scr.impl.BundleComponentActivator.initialEnable(BundleComponentActivator.java:305)
	at org.apache.felix.scr.impl.Activator.loadComponents(Activator.java:554)
	at org.apache.felix.scr.impl.Activator.access$200(Activator.java:70)
	at org.apache.felix.scr.impl.Activator$ScrExtension.start(Activator.java:421)
	at org.apache.felix.scr.impl.AbstractExtender.createExtension(AbstractExtender.java:196)
	at org.apache.felix.scr.impl.AbstractExtender.modifiedBundle(AbstractExtender.java:169)
	at org.apache.felix.scr.impl.AbstractExtender.modifiedBundle(AbstractExtender.java:49)
	at org.osgi.util.tracker.BundleTracker$Tracked.customizerModified(BundleTracker.java:488)
	at org.osgi.util.tracker.BundleTracker$Tracked.customizerModified(BundleTracker.java:1)
	at org.osgi.util.tracker.AbstractTracked.track(AbstractTracked.java:232)
	at org.osgi.util.tracker.BundleTracker$Tracked.bundleChanged(BundleTracker.java:450)
	at org.eclipse.osgi.internal.framework.BundleContextImpl.dispatchEvent(BundleContextImpl.java:944)
	at org.eclipse.osgi.framework.eventmgr.EventManager.dispatchEvent(EventManager.java:234)
	at org.eclipse.osgi.framework.eventmgr.ListenerQueue.dispatchEventSynchronous(ListenerQueue.java:151)
	at org.eclipse.osgi.internal.framework.EquinoxEventPublisher.publishBundleEventPrivileged(EquinoxEventPublisher.java:229)
	at org.eclipse.osgi.internal.framework.EquinoxEventPublisher.publishBundleEvent(EquinoxEventPublisher.java:138)
	at org.eclipse.osgi.internal.framework.EquinoxEventPublisher.publishBundleEvent(EquinoxEventPublisher.java:130)
	at org.eclipse.osgi.internal.framework.EquinoxContainerAdaptor.publishModuleEvent(EquinoxContainerAdaptor.java:217)
	at org.eclipse.osgi.container.Module.publishEvent(Module.java:499)
	at org.eclipse.osgi.container.Module.start(Module.java:486)
	at org.eclipse.osgi.container.ModuleContainer$ContainerStartLevel$2.run(ModuleContainer.java:1847)
	at org.eclipse.osgi.container.ModuleContainer$ContainerStartLevel$1.execute(ModuleContainer.java:1835)
	at org.eclipse.osgi.container.ModuleContainer$ContainerStartLevel.incStartLevel(ModuleContainer.java:1840)
	at org.eclipse.osgi.container.ModuleContainer$ContainerStartLevel.incStartLevel(ModuleContainer.java:1784)
	at org.eclipse.osgi.container.ModuleContainer$ContainerStartLevel.doContainerStartLevel(ModuleContainer.java:1745)
	at org.eclipse.osgi.container.ModuleContainer$ContainerStartLevel.dispatchEvent(ModuleContainer.java:1667)
	at org.eclipse.osgi.container.ModuleContainer$ContainerStartLevel.dispatchEvent(ModuleContainer.java:1)
	at org.eclipse.osgi.framework.eventmgr.EventManager.dispatchEvent(EventManager.java:234)
	at org.eclipse.osgi.framework.eventmgr.EventManager$EventThread.run(EventManager.java:345)

Dump of callerThis
null
```
.

**Comment by PierreFeillet** [comment]
Hi, Any news on this issue? Is there a plan to provide a random number generator that is FIPS certified in Q2? CP4BA has a deal at stake for Q2 that requires to be FIPS compliant with Semeru 11.

**Comment by PierreFeillet** [comment]
ODM bumped in the similar issue and solved it by changing the code as follows: 
`gen = new SecureRandom(); //SecureRandom.getInstance("SHA1PRNG");` 
Basically you should ask for a secure random generator but not for a specific algorithm. And we find one in Java 11 with FIPS activated.

**Comment by FrancoisTrible** [comment]
tracked in https://jsw.ibm.com/browse/DBACLD-85396

**Comment by lianxj-ibm** [comment]
Comments from CPE developer Roger Bacalzo: 
 That workaround does not apply to CPE, since the problem is being encountered at the Liberty OIDC client, not in any CPE code.  So there is no CPE code that can be changed that would allow it to specify a different random provider.  So we need to wait for a Liberty fix to resolve.
So it's a blocker for FNCM

**Comment by NottyCode** [comment]
The following files reference SHA1PRNG so they should also be addressed:

* dev/com.ibm.ws.security.openidconnect.clients.common/src/com/ibm/ws/security/openidconnect/jwk/Util.java
* dev/com.ibm.ws.jsf.2.2/src/org/apache/myfaces/application/viewstate/SessionIdGenerator.java
* dev/com.ibm.ws.security.common/src/com/ibm/ws/security/common/random/RandomUtils.java
* dev/com.ibm.ws.security.saml.websso.2.0/src/com/ibm/ws/security/saml/sso20/internal/utils/SamlUtil.java
* dev/com.ibm.ws.security.openidconnect.clients.common/src/com/ibm/ws/security/openidconnect/clients/common/OidcUtil.java
* dev/com.ibm.ws.security.oauth/src/com/ibm/oauth/core/internal/OAuthUtil.java
* dev/com.ibm.ws.security.social/src/com/ibm/ws/security/social/internal/utils/SocialUtil.java
* dev/com.ibm.ws.jsf.2.2/src/org/apache/myfaces/application/StateCache.java
* dev/com.ibm.ws.security.common/src/com/ibm/ws/security/common/jwk/utils/JsonUtils.java

**Comment by FrancoisTrible** [comment]
It seems #24763 was merged - when can we expect a liberty version with that fix ?

**Comment by arunavemulapalli** [comment]
I updated all sso classes 
jsf.2.2 classes look good

**Comment by FrancoisTrible** [comment]
Which liberty version will have that fix ?

**Comment by arunavemulapalli** [comment]
probably next fix pack version,  23.0.0.4
